editor/src/components/editor/persistence/persistence.spec.ts

```diff
@@ -0,0 +1,528 @@
+    expect(localProjects[localProjectKey(capturedData.newProjectId!)]!.model).toEqual(BaseModel)
+
+    testMachine.login()
+    await delay(20)
+
+    // Check that logging in uploaded it and deleted the local version
+    expect(mockSaveLog[capturedData.newProjectId!]).toEqual([BaseModel])
+    expect(localProjects[localProjectKey(capturedData.newProjectId!)]).toBeUndefined()
+
+    testMachine.save(ProjectName, FirstRevision, 'force')
+    await delay(20)
+    testMachine.stop()
+
+    // Check that future saves go to the server
+    expect(mockSaveLog[capturedData.newProjectId!]).toEqual([BaseModel, FirstRevision])
+    expect(localProjects[localProjectKey(capturedData.newProjectId!)]).toBeUndefined()
+  })
+
+  it('Logging out mid-session will switch to local saving', async () => {
+    const { capturedData, testMachine } = setupTest()
+
+    testMachine.login()
+    await delay(20)
+    testMachine.createNew(ProjectName, BaseModel)
+    await delay(20)
+
+    // Check it was saved to the server
+    expect(localProjects[localProjectKey(capturedData.newProjectId!)]).toBeUndefined()
+    expect(capturedData.newProjectId).toBeDefined()
+    expect(mockSaveLog[capturedData.newProjectId!]).toEqual([BaseModel])
+
+    testMachine.logout()
+    await delay(20)
+
+    testMachine.save(ProjectName, FirstRevision, 'force')
+    await delay(20)
+    testMachine.stop()
+
+    // Check it was saved locally
+    expect(localProjects[localProjectKey(capturedData.newProjectId!)]).toBeDefined()
+    expect(localProjects[localProjectKey(capturedData.newProjectId!)]!.model).toEqual(FirstRevision)
+    expect(mockSaveLog[capturedData.newProjectId!]).toEqual([BaseModel]) // Should be no new saves
+  })
+})
+
+describe('Loading a project', () => {
+  function addServerProject(id: string, model: PersistentModel) {
+    serverProjects[id] = model
+  }
+
+  function addLocalProject(id: string, model: PersistentModel) {
+    const now = new Date().toISOString()
+    localProjects[localProjectKey(id)] = {
+      model: model,
+      createdAt: now,
+      lastModified: now,
+      thumbnail: '',
+      name: ProjectName,
+    }
+  }
+
+  it('Loads a server project', async () => {
+    const { capturedData, testMachine } = setupTest()
+
+    const projectId = randomProjectID()
+    addServerProject(projectId, BaseModel)
+
+    testMachine.load(projectId)
+    await delay(20)
+
+    expect(capturedData.projectNotFound).toBeFalsy()
+    expect(capturedData.createdOrLoadedProject).toEqual(BaseModel)
+  })
+
+  it('Loads a local project', async () => {
+    const { capturedData, testMachine } = setupTest()
+
+    const projectId = randomProjectID()
+    addLocalProject(projectId, BaseModel)
+
+    testMachine.load(projectId)
+    await delay(20)
+
+    expect(capturedData.projectNotFound).toBeFalsy()
+    expect(capturedData.createdOrLoadedProject).toEqual(BaseModel)
+  })
+
+  it('Favours a local project over a server project', async () => {
```

Comment:
does it mean if a project exists both locally and on the server, we favor the local version? is this right? is this because we want to handle the editor going temporarily offline?

Reply:
Yeah exactly that. Every server save of a project ensures that the local version of it is deleted.

Reply:
(This mimics the current persistence behaviour btw)